src/main/python/systemml/defmatrix.py

```diff
@@ -555,66 +620,304 @@ def __repr__(self):
         """
         This function helps to debug matrix class and also examine the generated PyDML script
         """
-        if self.data is None:
-            print('# This matrix (' + self.ID + ') is backed by below given PyDML script (which is not yet evaluated). To fetch the data of this matrix, invoke toNumPyArray() or toDataFrame() or toPandas() methods.\n' + eval([self], execute=False))
-        elif isinstance(self.data, DataFrame):
-            print('# This matrix (' + self.ID + ') is backed by PySpark DataFrame. To fetch the DataFrame, invoke toDataFrame() method.')
+        if self.eval_data is None:
+            print('# This matrix (' + self.ID + ') is backed by below given PyDML script (which is not yet evaluated). To fetch the data of this matrix, invoke toNumPy() or toDF() or toPandas() methods.\n' + eval([self], execute=False))
         else:
-            print('# This matrix (' + self.ID + ') is backed by NumPy array. To fetch the NumPy array, invoke toNumPyArray() method.')
+            print('# This matrix (' + self.ID + ') is backed by ' + str(type(self.eval_data)) + '. To fetch the DataFrame or NumPy array, invoke toDF() or toNumPy() method respectively.')
         return ''
+    
+    ######################### NumPy related methods ######################################
+    
+    __array_priority__ = 10.2
+    ndim = 2
+    
+    THROW_ARRAY_CONVERSION_ERROR = False
+    
+    def __array__(self, dtype=np.double):
+        """
+        As per NumPy from Python,
+        This method is called to obtain an ndarray object when needed. You should always guarantee this returns an actual ndarray object.
+        
+        Using this method, you get back a ndarray object, and subsequent operations on the returned ndarray object will be singlenode.
+        """
+        if not isinstance(self.eval_data, SUPPORTED_TYPES):
+            # Only warn if there is an unevaluated operation (which could potentially generate large matrix or if data is non-supported singlenode formats)
+            import inspect
+            frame,filename,line_number,function_name,lines,index = inspect.stack()[1]
+            msg = 'Conversion from SystemML matrix to NumPy array (occurs in ' + str(filename) + ':' + str(line_number) + ' ' + function_name + ")"
+            if matrix.THROW_ARRAY_CONVERSION_ERROR:
+                raise Exception('[ERROR]:' + msg)
+            else:
+                print '[WARN]:' + msg
```

Comment:
Can you update this to Python 3 compatible syntax, as this currently throws an error as soon as the `systemml` package is imported.